src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+
+                        if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+
+                        Debug.Assert(CurrentToken.Kind == SyntaxKind.CommaToken);
+                        types.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                    }
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.IsTerminator(),
+                    expected: SyntaxKind.CommaToken);
+            }
+        }
+
+        private bool IsFunctionPointerParameterModifier()
+            => IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword;
+
+        private bool IsFunctionPointerStart()
+            => CurrentToken.Kind == SyntaxKind.DelegateKeyword && PeekToken(1).Kind == SyntaxKind.AsteriskToken;
+
+        // We attempt to make error recovery a little nicer here by only treating a single
+        // identifier token as a calling convention if the token afterwords is actually a
+        // possible start to the function pointer parameter list. The intuition is that, for
+        // some partial trees like this:
+        //
+        //  (delegate*MyProperty.MyMethod();
+        //
+        // The user is actually in the middle of adding a cast, and if we interpreted MyProperty
+        // as a calling convention, we'd get a much worse set of errors than if we treated the
+        // function pointer as having no calling convention.
+        // Also for nicer error recovery, we only treat predefined types as if they could be the
+        // calling convention. For any other keyword, they're almost certainly part of some other
+        // construct, and would produce better errors if parsed separately. The user could have this:
+        //
+        // delegate* // Declaration in progress
+        // while (true) {}
```

Comment:
>while (true) {} [](start = 11, length = 15)

I think the same argument could be made for predefined types:
```
delegate* // Declaration in progress
int x = 0;
```
Perhaps we should eat any keyword as long as it is on the same line? Which would mean that a predefined type on a different line is not a calling convention.

Reply:
That's a fair point, though I do think that, for things like `foreach`, `while`, `if`, etc, it would never be a good idea to eat them here, since they are control flow concepts and could never be thought to have meaning here. I'll look into preventing keyword eating when it's not on the same line, though I'll also need to test that we eat a regular identifier on another line there.

---
In reply to: [349363152](https://github.com/dotnet/roslyn/pull/39879#discussion_r349363152) [](ancestors = 349363152)

Reply:
i like this idea. 👍